test/840_weave_kube_3_test.sh

```diff
@@ -131,30 +120,62 @@ spec:
   selector:
     run: nettest
 EOF
+}
+
+function setup_kubernetes_cluster {
+    greyly echo "Setting up kubernetes cluster"
+    tear_down_kubeadm;
+
+    # Make an ipset, so we can check it doesn't get wiped out by Weave Net
+    docker_on $HOST1 run --rm --privileged --net=host --entrypoint=/usr/sbin/ipset weaveworks/weave-npc create test_840_ipset bitmap:ip range 192.168.1.0/24 || true
+    docker_on $HOST1 run --rm --privileged --net=host --entrypoint=/usr/sbin/ipset weaveworks/weave-npc add test_840_ipset 192.168.1.11
 
-podName=$($SSH $HOST1 "$KUBECTL get pods -l run=nettest -o go-template='{{(index .items 0).metadata.name}}'")
+    # kubeadm init upgrades to latest Kubernetes version by default, therefore we try to lock the version using the below option:
+    k8s_version="$(run_on $HOST1 "kubelet --version" | grep -oP "(?<=Kubernetes )v[\d\.\-beta]+")"
+    k8s_version_option="$([[ "$k8s_version" > "v1.6" ]] && echo "kubernetes-version" || echo "use-kubernetes-version")"
 
-check_all_pods_communicate() {
-    if [ -n podIP ] ; then
-        status=$($SSH $HOST1 "$KUBECTL exec $podName -- curl -s -S http://127.0.0.1:8080/status")
-        if [ $status = "pass" ] ; then
-            return 0
+    for host in $HOSTS; do
+        if [ $host = $HOST1 ] ; then
+        run_on $host "sudo systemctl start kubelet && sudo kubeadm init --$k8s_version_option=$k8s_version --token=$TOKEN"
+        else
+        run_on $host "sudo systemctl start kubelet && sudo kubeadm join --token=$TOKEN $HOST1IP:$KUBE_PORT"
         fi
-    fi
-    return 1
+    done
+
+    # Ensure Kubernetes uses locally built container images and inject code coverage environment variable (or do nothing depending on $COVERAGE):
+    sed -e "s%imagePullPolicy: Always%imagePullPolicy: Never%" \
+        -e "s%env:%$COVERAGE_ARGS%" \
+        "$(dirname "$0")/../prog/weave-kube/weave-daemonset-k8s-1.6.yaml" | run_on "$HOST1" "$KUBECTL apply -n kube-system -f -"
+}
+
+function teardown_kubernetes_cluster {
+    greyly echo "Tearing down kubernetes cluster"
+    tear_down_kubeadm; 
+
+    # Destroy our test ipset, and implicitly check it is still there
+    assert_raises "docker_on $HOST1 run --rm --privileged --net=host --entrypoint=/usr/sbin/ipset weaveworks/weave-npc destroy test_840_ipset"
+
 }
 
-assert_raises 'wait_for_x check_all_pods_communicate pods'
+function cleanup {
+teardown_kubernetes_cluster
+}
+trap cleanup EXIT
 
-# Check that a pod can contact the outside world
-assert_raises "$SSH $HOST1 $KUBECTL exec $podName -- $PING 8.8.8.8"
+function main {
```

Comment:
Maybe it's a personal taste, but I prefer to have assertions in locations as they used to be (unless they are repeated). Having them in the functions requires a bit more context switching
when reading the suite.

Also, IMHO, `main` is a bit an overkill.